### PR TITLE
router: add per_try_idle_timeout configuration

### PR DESCRIPTION
**Pull request by mattklein123**
Allows ensuring continual progress of individual request attempts.

Risk Level: Low when not using feature. Medium when using feature. Timeouts are scary.
Testing: Unit and integration.
Docs Changes: Added
Release Notes: Added
Platform Specific Features: N/A

### CODE REVIEW

File: source/common/router/upstream_request.cc [hunk 1]
```diff
@@ -331,13 +337,32 @@ void UpstreamRequest::resetStream() {
...
+  if (parent_.timeout().per_try_idle_timeout_.count() > 0) {
+    per_try_idle_timeout_ =
+        parent_.callbacks()->dispatcher().createTimer([this]() -> void { onPerTryIdleTimeout(); });
+    resetPerTryIdleTimer();
+  }
+}
+
+void UpstreamRequest::onPerTryIdleTimeout() {
+  ENVOY_STREAM_LOG(debug, "upstream per try idle timeout", *parent_.callbacks());
+  stream_info_.setResponseFlag(StreamInfo::ResponseFlag::StreamIdleTimeout);
```
Comment (alyssawilk) [inline source/common/router/upstream_request.cc:364]: If there's 2 tries (hedging or retries) but we successfully proxy, should the response flag be set?

**Review by alyssawilk — COMMENTED** [review]
Looks good overall!  Just 2 nits/questions

File: api/envoy/config/route/v3/route_components.proto [hunk 2]
```diff
@@ -1318,6 +1318,27 @@ message RetryPolicy {
   //   would have been exhausted.
   google.protobuf.Duration per_try_timeout = 3;
 
+  // Specifies an upstream idle timeout per retry attempt (including the initial attempt). This
+  // parameter is optional and if absent there is no per try idle timeout. The semantics of the per
+  // try idle timeout are similar to the
+  // :ref:`route idle timeout <envoy_v3_api_field_config.route.v3.RouteAction.timeout>` and
+  // :ref:`stream idle timeout
+  // <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.stream_idle_timeout>`
+  // both enforced by the HTTP connection manager. The differences is that this idle timeout
```
Comment (adisuissa) [inline api/envoy/config/route/v3/route_components.proto:1327]:
```suggestion
  // both enforced by the HTTP connection manager. The difference is that this idle timeout
```

File: api/envoy/config/route/v3/route_components.proto [hunk 3]
```diff
@@ -1318,6 +1318,27 @@ message RetryPolicy {
...
+  // is useful in cases in which total request timeout is bounded by a number of retries and a
+  // :ref:`per_try_timeout <envoy_v3_api_field_config.route.v3.RetryPolicy.per_try_timeout>`, but
+  // there is a desire to ensure each try is making incremental progress. Note also that similar
+  // to :ref:`per_try_timeout <envoy_v3_api_field_config.route.v3.RetryPolicy.per_try_timeout>`,
+  // this idle timeout does not start until after both the entire request has been received by the
+  // router *and* a connection pool connection has been obtained. Unlike
+  // :ref:`per_try_timeout <envoy_v3_api_field_config.route.v3.RetryPolicy.per_try_timeout>`,
+  // the idle timer continues once the response starts streaming back to the downstream client.
+  // This ensures that response data continues to make progress when not also using one of the
+  // HTTP connection manager idle timeouts.
```
Comment (adisuissa) [inline api/envoy/config/route/v3/route_components.proto:1339]: I don't fully understand this.
Do you mean to say that:
```
This ensures that response data continues to make progress without using one of the HTTP connection manager idle timeouts.
```
or
```
This ensures that response data continues to make progress when none of the HTTP connection manager idle timeouts are used.
```
(Assuming the former, but not sure).

**Review by adisuissa — COMMENTED** [review]
Thanks!
Left a couple of API comments

File: source/common/router/upstream_request.cc (hunk 1 above)
Comment (mattklein123) [reply source/common/router/upstream_request.cc:364]: This is stream info on the upstream request (for upstream logs), not the downstream stream info.

File: envoy/stream_info/stream_info.h [hunk 4]
```diff
@@ -158,6 +158,8 @@ struct ResponseCodeDetailValues {
   const std::string UpstreamTimeout = "upstream_response_timeout";
   // The final upstream try timed out.
   const std::string UpstreamPerTryTimeout = "upstream_per_try_timeout";
+  // The final upstream try idle timed out.
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply envoy/stream_info/stream_info.h:161]: Sorry update to what? I added "idle?"

[the remark this answers is not in the document]
Comment (alyssawilk) [reply envoy/stream_info/stream_info.h:161]: oh sorry PEBKAC

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**